torch/csrc/jit/passes/common_subexpression_elimination.cpp

```diff
@@ -0,0 +1,145 @@
+// Do NOT support t/ts/g/gs attributes.
+// If t/ts are supported, CONSTANT node comparison may need to consider device.
+bool attributesEqualCSE(const Node* lhs, const Node* rhs) {
+  JIT_ASSERT(lhs != nullptr);
+  JIT_ASSERT(rhs != nullptr);
+  // One has attributes, the other does not.
+  if (lhs->hasAttributes() != rhs->hasAttributes()) return false;
+  // Neither has attributes.
+  if (!lhs->hasAttributes() && !rhs->hasAttributes()) return true;
+
+  auto lnames = lhs->attributeNames();
+  auto rnames = rhs->attributeNames();
+  if (lnames != rnames) return false;
+
+  Node* l = const_cast<Node*>(lhs);
+  Node* r = const_cast<Node*>(rhs);
+  for (auto name : lnames) {
+    if (l->kindOf(name) != r->kindOf(name)) return false;
+
+    #define COMPARE_ATTRIBUTEVALUE(type) \
+      case AttributeKind::type: \
+        { if (l->type(name) != r->type(name)) return false; } break;
+
+    switch(l->kindOf(name)) {
+      COMPARE_ATTRIBUTEVALUE(f)
+      COMPARE_ATTRIBUTEVALUE(fs)
+      COMPARE_ATTRIBUTEVALUE(i)
+      COMPARE_ATTRIBUTEVALUE(is)
+      COMPARE_ATTRIBUTEVALUE(s)
+      COMPARE_ATTRIBUTEVALUE(ss)
+      default:
+        return false;
+    }
+
+    #undef COMPARE_ATTRIBUTEVALUE
+  }
+
+  return true;
+}
+
+// Later, if someone wants to reuse this, it can be moved to some header files.
+inline void hash_combine(size_t& seed, size_t value) {
+  seed ^= value + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+}
+
+struct HashNodeCSE {
+  size_t operator()(const Node* k) const {
+    JIT_ASSERT(k != nullptr);
+    size_t seed = 0;
+    hash_combine(seed, k->kind());
+    hash_combine(seed, k->stage());
+    for (auto i : k->inputs()) {
+      hash_combine(seed, i->unique());
+    }
+    return seed;
+  }
+};
+
+struct EqualNodeCSE {
+  bool operator()(const Node* lhs, const Node* rhs) const {
+    if (lhs == nullptr && rhs == nullptr) return true;
+    if (lhs == nullptr || rhs == nullptr) return false;
+
+    // Check whether two nodes are the same kind.
+    if (lhs->kind() != rhs->kind()) return false;
+
+    // Check the stage.
+    if (lhs->stage() != rhs->stage()) return false;
+
+    // Check whether the inputs are the same.
+    if (lhs->inputs().size() != rhs->inputs().size()) return false;
+
+    if (!std::equal(lhs->inputs().begin(), lhs->inputs().end(), rhs->inputs().begin())) return false;
+
+    // Check the attributes.
+    if (!attributesEqualCSE(lhs, rhs)) return false;
+
+    return true;
+  }
+};
+
+// The function implements common subexpression elimination.
+// Since the nodes are visited in topological order, one pass is enough.
+void EliminateCommonSubexpression(std::shared_ptr<Graph>& graph) {
+  auto nodes = graph->nodes();
+  std::unordered_set<Node*, HashNodeCSE, EqualNodeCSE> subexprs;
+  for (auto it = nodes.begin(); it != nodes.end(); ++ it) {
+    auto node = *it;
+    if (node->kind() == kPythonOp
+        || node->kind() == kCppOp
+        || node->kind() == kEval
+        || node->kind() == kUndefined
```

Comment:
I think, technically, Undefined nodes should be OK, because they actually just correspond to "None" inputs (i.e. null variables, i.e., undefined ATen tensors.)

Reply:
I see. Will remove Undefined nodes from black list.